### ISSUE TITLE
change wording when reporting to "expected" and "actual"

### ISSUE THREAD

**Issue by igorakkerman** [robstoll/atrium-roadmap#55]
I keep being confused whenever I see a new assertion error like:
```
expect: "report"        <753332884>
◆ to be: "easy to understand"        <1978033453>
◆ to start with "easy"
```
The main reason for this is: 
- the _actual_ result comes after the word `expect`
- the _expected_ result is not preceded by the word `should`

In the example above, at first glance, it seems like `"report"` was expected but instead the actual value is `"easy to understand"`. While `expect("report").toBe("easy to understand")` reads great in the code, my personal sensation is confusion when it comes to the report.

How about having clear, direct comparisons:
```   
◆ should be equal:
    expected: "easy to understand"
    actual:   "report"      
◆ should start with: 
    expected: "easy"
    actual:   "report"
```

Btw., this style would also make complex differences in more complex data class objects more obvious and easier to detect.

**Comment by robstoll** [comment]
Can you please fix the sentence
> In the example above [...] actual value is `"clearer"` 

there is no "clearer" in the example.
And it would be nice if you could elaborate on the 
> would also make complex differences [...] easier to detect

An example would be good.
A bit background for the current way of reporting, there are several intends:
- reflect the way one writes assertions with both API styles; it would not fit a `"report" shouldBe "easy to understand"` style, but we don't have this one (yet)
- the notation is a bit more mathematical like (I don't want to claim it is mathematical :D) and `expect` could be replaced with `it is valid for`

**Comment by jGleitz** [comment]
I, personally, was at first also irritated by atrium’s reporting. However, I now see and like the logic behind it and would not change it fundamentally. But we can think about what would make it easier to understand what is going on.

The main issue I have nowadays is that I still sometimes need more than a few milliseconds to see what is actually wrong. With hamcrest, I could always search for the word “but” and would directly see the problem (in general, I think hamcrest’s reporting is worse than artium’s, though). I think https://github.com/robstoll/atrium-roadmap/issues/27 could bring big improvements in that regard.

I think @igorakkerman’s confusion stems in part from the following: A user who is not familiar with atrium sees the phrase “expect: xyz” and interprets it first as meaning “we expected to see the value xyz”. Then the user reads the following lines, sees that they make no sense for the current interpretation and has to backtrack to the beginning to make sense of it all. At least, that is the process I went through in the beginning.
Can we think of a way to rephrase the current reporting to avoid that trap, without changing it fundamentally?

I will start with a suggestion, even though I do not like it myself that much:

```
failure for subject: "report"        <753332884>
◆ expected to be: "easy to understand"        <1978033453>
◆ expected to start with "easy"
```

**Comment by robstoll** [comment]
I first promoted to use `assert` as assertion verb and also used it in reporting. It was due to bugs in intellij that I switched to `expect`. Maybe reporting would be clearer if we used `assert` again (also in case someone uses `expect` as verb?).

Note that the solution should also work if someone reports not only failing assertions.

**Comment by jGleitz** [comment]
I think using “assert” would also help, although my version is clearer, albeit more verbose. @igorakkerman?

> Note that the solution should also work if someone reports not only failing assertions.

Would also work for my solution, one would of course have to omit the “failure for” if there is no failure.

**Comment by igorakkerman** [comment]
I really like the version @jGleitz suggests, better than mine. In this case, verbosity helps. What about this attempt?
```
Assertion failed. 
❗❗ Actual:
    "report"        <753332884>
◆ expected to be: 
    "easy to understand"        <1978033453>
◆ expected to start with: 
    "easy"
```
By differences in complex objects, I meant that in real applications we mostly compare business objects with many properties and the error often hides in a single character or digit. The difference is easier to spot if the String representations are displayed horizontally aligned:
```
Assertion failed. 
❗❗ Actual:
    FamilyMember(name=MyDog, age=10.5)        (FamilyMember <420421290>)
◆ expected to be: 
    FamilyMember(name=MyD0g, age=18.5)        (FamilyMember <1251969156>)
```

(in the example, there are two differences)

**Comment by jGleitz** [comment]
Hm, having the values on a new line will make atrium’s output have _even more_ more lines. It can already get quite long and that would mean almost doubling the number of lines!

On the other hand, it would make aligning values a lot easier (https://github.com/robstoll/atrium/issues/295). And on top, objects with newlines in their `toString` would not break the reporting that badly anymore (I do not know how common that is, though).

**Comment by jGleitz** [comment]
I might add that if diff’ing the actual with the expected is the main concern, providing a diff, either in the output itself or through the IDE (https://github.com/robstoll/atrium-roadmap/issues/39) might address the issue better.

**Comment by igorakkerman** [comment]
I believe that a diff like the one in the referenced issue would rather make the output harder to read:
```
>> diff: 
    @@ -1 +1 @@
    [-Hello-]{+hello+} my name is Robert and I am the author of [-Atrium-]{+ATrium+}
```
I believe that what matters most is productivity. Whenever I see a diff, I have to start thinking even more. The output should be "visually clean". Then the failing issue will be identified the fastest.

**Comment by robstoll** [comment]
I would argue that you only have to know how to read the diff view if you are not yet familiar with and are much faster afterwards (and it is not hard to understand it IMO)
Spotting the error of ATrium instead of Atrium is much harder with the following compared to the diff IMO as the diff already gives visual help what can be skipped over without hesitation and one does not have to scan each letter one by one.
```
expect:  "hello my name is Robert and I am the author of ATrium"
to be:   "Hello my name is Robert and I am the author of Atrium"
```
 It gets worse if the text is multi-line. With a diff view I would simply skip lines which are equal. 

And I want to point out three things in addition. 
1. one can already use multi-line reporting (see https://github.com/robstoll/atrium#reporterbuilder)
2. improvements for comparing data classes are already planned: https://github.com/robstoll/atrium/issues/151
3. Atrium suffers from verbosity issues in conjunction with feature assertions. That's something I planned for 0.11.0 as they are in Atrium since 0.6.0 and no one opened an issue so far (or in other words, it seems the bug does not really bother). We can try to tackle them in 0.10.0 instead but I still have the feeling this should have a low priority. 

Last but not least, I ask myself (well I have quite a biased view) if it is really that hard to understand the reporting after reading the explanation in:
- https://github.com/robstoll/atrium#your-first-assertion
- and for property/method assertions: https://github.com/robstoll/atrium#property-and-methods

I have the feeling it is all there, e.g. 
> a feature starts with a ▶ followed by the feature's name and its actual value. 

Maybe it would be enough to improve the docs?

**Comment by robstoll** [comment]
Maybe it would help if we use `is equal to` in reporting?

Please consider for your ideas that assertion functions can also be used in contains assertions to identify an entry. For instance:
```
expect: [1, 2, 2, 4]        (java.util.Arrays.ArrayList <1234789>)
◆ contains, in any order: 
  ⚬ an entry which: 
      » is less than: 0        (kotlin.Int <1234789>)
    ⚬ ▶ number of occurrences: 0
        ◾ is at least: 1
```
The `is less than: 0` is also just reporting the assertion, having:
```
  ⚬ an entry which: 
      » expected to be less than: 0        (kotlin.Int <1234789>)
```
is not nice IMO and I definitely don't want to have different reporting of the same assertion in different contexts as default case. This would make the reporting even more complex than it already is.

**Comment by jGleitz** [comment]
Regarding:

 > one can already use multi-line reporting

> Maybe it would be enough to improve the docs?

I beg to differ. I think atrium should have the goal to be as easy to understand for the broadest possible audience _per default_. It is of course very difficult to find out what “easy to understand for many people” means and that is why we are having this discussion. However, I do not like the notion of “it does not matter that much, people can tweak it if they do not like it” as long as we are not sure that we have sensible defaults.

I, for example, have used atrium a lot by now and tried to read most of the docs. I still was not aware that I can turn on multi-line reporting.

@robstoll Could we add “the default settings for reporting are according to what the contributor’s think is clearest and easiest to read for the broadest scope of peoble possible” (or something like that) to the [requirements](https://github.com/robstoll/atrium-roadmap/wiki/Requirements)?

**Comment by jGleitz** [comment]
> Maybe it would help if we use is equal to in reporting?

I do not think so. As I tried to explain, I misunderstood the reporting because my brain took a wrong turn in the first line. The next lines made my mistake clear to me. So it is not about them.

**Comment by igorakkerman** [comment]
1. `is equal to` is much better than `is`, I also wanted to suggest it.
2. Verbosity in this case improves productivity, so embrace verbosity.
3. For a single line, diffs almost always take longer to read than a visual comparison. (I have been reading diffs for 20 years now.) In reports, IMO they produce visual noise.
4. Multi-line String comparison is an edge case that could be handled differently but the propos ed solution at least doesn't make it worse.
5. Reading and understanding the report should be intuitive and come natural and not require studying the documentation first and "learning to read the report". 
6. The expected report output should guide the framework design. Not the other way around.

My proposal for your example above:
```
Assertion failed. 
❗❗ Actual: [1, 2, 2, 4]        (java.util.Arrays.ArrayList <1234789>)
◆ expected to contain, in any order: 
  ⚬ an item which:
      »  is less than: 0        (kotlin.Int <1234789>)
    ⚬ ▶ number of such items:
        ◾ Actual:   0
        ◾ Expected: at least 1
```

**Comment by jGleitz** [comment]
To chime in, because I wanted to write the same:

I do not yet know whether I like the proposal, but to clarify the idea:

I would write the “expected” only at the top level. At the top level, we always know what the start of the sentence was and can thus make sure that it makes sense. @igorakkerman gave the example.

**Comment by igorakkerman** [comment]
I thought about what the assertion actually means. The author doesn't care about the cardinality of the item. All that matters is: We wanted an item, we didn't get it. Hence:
```
Assertion failed. 
❗❗ Actual: [1, 2, 2, 4]        (java.util.Arrays.ArrayList <1234789>)
◆ expected to contain, in any order: 
  ⚬ an item which:
      »  is less than: 0        (kotlin.Int <1234789>)
◾ But no such item was found.
```

**Comment by robstoll** [comment]
> I beg to differ [...] that is why we are having this discussion. However, I do not like the notion of “it does not matter that much, people can tweak it if they do not like it” as long as we are not sure that we have sensible defaults.

I grant your beg :laughing: but you clearly got me wrong here and I did not write something like that; you make it sound like I would like to turn down this discussion. Not at all. I only wanted to share knowledge about Atrium - we could also change the default to new line instead of same line. And I agree that we should have good defaults, the one we have is already _sensible_ though, but maybe we can improve it ;) Let us stick to a constructive feedback here, otherwise the discussion becomes exhausting.

> “the default settings for reporting are according to what the contributor’s think is clearest and easiest to read for the broadest scope of peoble possible”

I guess you did not want to put it like a rule but it sound like one to me -- and I am afraid but I don't want to put rules in the requirements. We can put something about reporting but should be a bit more precise on what we want to achieve. So first, we should think about who is our average persona, newcomers, long term users or library authors. I assume you wanted to stress that it should be best for newcomers but maybe I am wrong?

The current reporting embraces the principle that it looks more or less like the code. Thus I would claim that someone new to Atrium and not spoiled by another assertion library would immediately know what the error states because it is what they just wrote, just substituting the unknown subject with its actual value. As far as I know we are all not native speakers, therefore let us have a look at the German reporting quickly `erwarte: 10 ist 9`, I guess I am too biased and look at it from a more mathematical angle than you two -- I guess you always put a `aber` in there `erwarte 10 aber ist 9` which of course makes it look like 10 is the actual value.  
Anyway, most users have most likely used another assertion library before and -- as I already showed -- I understand that you both confuse 10 being the expected value and 9 the actual. You also learned it this way via AssertJ or Hamcrest which both show the `expected` value first (same for Truth btw.)  

My conclusion is not that you have to learn something else but that we should stress that the first value we see is the subject (as you both suggested). I prefer to use `subject` instead of `actual` and hope that's fine for Igor as well. 

I suggest we don't try to change everything at once and focus on one thing in this issue (too many topics here IMO). How about we change the reporting to:
```
expect subject: "report"        <753332884>
◆ to be: "easy to understand"        <1978033453>
◆ starts with:  "easy"
```
This way we still are near to the code (another suggestion follows further down below).

> I would write the “expected” only at the top level. 

I prefer if we look for another idea as it requires 
> reporting of the same assertion in different contexts

and as I already stated

> I definitely don't want to have different reporting of the same assertion in different contexts as default case

We can do it exceptionally but IMO we should have a solution which does not require it in general.
@jGleitz I hope I have covered all your concerns and points.


Let's have a quick look at @igorakkerman's points

1. `is equal to` is much better than `is`, I also wanted to suggest it. 
   I am kind of in favour of changing `to be` to `is equal to` as some confuse it with `is same instance  as` but then reporting does not reflect the code as closely any more. If we decide to use `is equal to` then I would suggest that we write: 
   ```
   expect that subject: "report"        <753332884>
   ◆ is equal to: "easy to understand"        <1978033453>
   ◆ starts with:  "easy"
   ```
   @jGleitz fine with you as well? or are you against it?

   We could even think about adding ✘ at the front to make it clear that we are reporting failures (even though the default is only failure reporting).
   ```
   expect that subject: "report"        <753332884>
   ✘ is equal to: "easy to understand"        <1978033453>
   ✘ starts with:  "easy"
   ```

2. Verbosity in this case improves productivity, so embrace verbosity. 
  you are talking about the diff right? I think we should move this discussion to #39. I have nothing against verbosity in general, it should just not become repetitive.

3. For a single line, diffs almost always take longer to read than a visual comparison. (I have been reading diffs for 20 years now.) In reports, IMO they produce visual noise.
  Added your view to #39, leave a comment there in case I misunderstood you
4. Multi-line String comparison is an edge case that could be handled differently but the propos ed solution at least doesn't make it worse.
   You mean because of the alignment? 
5. Reading and understanding the report should be intuitive and come natural and not require studying the documentation first and "learning to read the report". 
   I agree and don't think that one needs to read the docs to understand `expect: 10 to be 9` but you have proven me wrong :wink: 

6. The expected report output should guide the framework design. Not the other way around.
  Very general statement and I don't know what exactly you refer to. I guess the statement I made like "think about reporting not only failures". Or "think about that assertion functions can be used to identify entries" Those stem from requirements and not from the implementation. If you meant something different then feel free to elaborate.

**Comment by jGleitz** [comment]
First of all: sorry for misunderstanding you, it certainly was not intentional. If you felt that I was not being constructive, please excuse! My only intention was to convey what I think is important for atrium.

<hr>

> I don't want to put rules in the requirements

I wanted to formulate a goal, not a rule. But I see from your reaction that my formulation does not convey to you what I wanted it to convey.

> I assume you wanted to stress that it should be best for newcomers but maybe I am wrong?

No, that is exactly what I did not want. I wanted to express that we should try to balance the needs of _all_ user groups. This is what I meant by “or the broadest scope of people possible”


I understand your points against the suggestions we brought up until now. “`expect subject`” will already be a good improvement. Maybe it even is the best compromise.

<hr>

> the principle that it looks more or less like the code

You bring that up a lot. I actually thought we dropped that requirement after https://github.com/robstoll/atrium-roadmap/issues/53. If you still think this should be a goal, can we discuss it (maybe in a new ticket)? Because I actually do not understand your reasons for that goal. 

<hr>

> @jGleitz [is `is equal to`] fine with you as well? or are you against it?

I have no opinion on this. Both are equally fine for me.

<hr>

> We could even think about adding ✘ at the front to make it clear that we are reporting failures (even though the default is only failure reporting).

I am in favour. I think it is a clever and easy trick to make the reporting easier to understand.

**Comment by robstoll** [comment]
> No, that is exactly what I did not want. I wanted to express that we should try to balance the needs of all user groups. This is what I meant by “or the broadest scope of people possible”

I see, go ahead and try to formulate a requirement please. 

> You bring that up a lot. I actually thought we dropped that requirement after #53. If you still think this should be a goal, can we discuss it (maybe in a new ticket)? Because I actually do not understand your reasons for that goal.

I don't see it as a goal. It's a principle which is applied now and not following the principle in certain cases makes reporting less intuitive in others. Which does not mean that we have to stick to it. I only wanted to make us aware of it. The more I think about it the more I am sure that `is equal to` is better than `to be` in reporting and there are surely other cases where not following is better. And the more we do this the more we can drop this principle entirely as it only makes sense if the user has the feeling that she sees in reporting what she just wrote. Moving the size check out of contains, does not reflect the code but only in a slight way IMO.
Please open another issue if you would like to discuss more.

created three tasks for this issue
- https://github.com/robstoll/atrium/issues/307
- https://github.com/robstoll/atrium/issues/308
- https://github.com/robstoll/atrium/issues/309

@igorakkerman I doubt that you have objections about using ✘ instead of bullet points and created already a corresponding issue. Open another issue if you would like to discuss it

**Comment by robstoll** [comment]
Thanks for the discussion, we found a few points which can be improved in Atrium. Please open issues if there are other things confusing you or if you have suggestions how Atrium could be improved.
